make prodigy command respect display-buffer-alist

lem102 opened the pull request:
closes #146

DamienCassou reviewed (commented):
Thank you for the PR. Can you please check the tests? They don't pass on the CI.

lem102 commented:
I'm struggling to run the tests locally. Are the docs for running the tests up to date? Running `make` doesn't seem to do anything related to running the tests.

DamienCassou commented:
> I'm struggling to run the tests locally.

you can always refer to how GitHub actions are implemented to learn how this works. This is in `.github/workflows/test.yml`:

```shellsession
$ make ci-dependencies
$ nix shell \
  --extra-experimental-features flakes \
  --extra-experimental-features nix-command \
  nixpkgs#nodePackages.coffee-script \
  --command make check
```

The Makefile is implemented with [makel](https://github.com/DamienCassou/makel). Reading its documentation may help understand how things work.

The `make ci-dependencies` downloads dependencies (look at the first line of the Makefile) if they are not checked out in the parent folder already.

Then, the `nix shell` command downloads coffee-script before executing `make check`. The test that fails is in `features/quit.feature` and this doesn't need coffee-script. As a result, you can replace the whole `nix shell` command with just:

```shellsession
$ make test-ecukes
```

DamienCassou commented:
I think the overall issue with this PR is that the `quit.feature` test suite requires `prodigy-mode` to only be executed once (to keep the service marks when reopening prodigy) whereas the PR executes `prodigy-mode` each time the user re-opens the prodigy buffer.

The PR can be fixed with:

```diff
diff --git a/prodigy.el b/prodigy.el
index e9bb211..07d1a57 100644
--- a/prodigy.el
+++ b/prodigy.el
@@ -1550,9 +1550,10 @@ (define-derived-mode prodigy-view-mode special-mode "Prodigy-view"
 (defun prodigy ()
   "Manage external services from within Emacs."
   (interactive)
-  (let ((buffer (get-buffer-create prodigy-buffer-name)))
+  (let ((buffer-p (prodigy-buffer))
+        (buffer (get-buffer-create prodigy-buffer-name)))
     (with-current-buffer buffer
-      (prodigy-mode)
+      (when (not buffer-p) (prodigy-mode))
       (prodigy-start-status-check-timer))
     (pop-to-buffer buffer)))
 
-- 
2.47.0
```

What still needs to be done is to add a new test scenario covering the new behavior.

lem102 commented:
Thanks Damien, I'm not super familiar with github actions, so thanks for the tip!

I'll look into adding a new test scenario when I have the time.

DamienCassou commented:
Hi @lem102. Do you still plan to work on this?

lem102 commented:
> Hi @lem102. Do you still plan to work on this?

Hi Damien,

Fixed the tests and added a new test to check display buffer rules are respected.

Let me know if there's anything else that needs doing :)

DamienCassou commented on features/prodigy.feature:
```diff
@@ -33,3 +33,12 @@ Feature: Prodigy
       | foo  | nil         |
       | qux  | nil         |
+
+  Scenario: Prodigy mode
```
we should be a little more explicit here. What is the scenario that this test is checking?

DamienCassou reviewed (changes requested):
Can you please rename the test scenario and squash all your commits into one? Thank you for your work.

DamienCassou reviewed (approved):
excellent work. Thank you very much.
